src/chunk_store.c

```diff
@@ -119,6 +118,11 @@ void chunk_store_take_node(node_t *node) {
     // This is a brand new node, so keep it!
     map_set(&seen_chunks, node_hash, node);
 
+    // NOTE: If this is a terminal node (node->id == 0), we *could*
+    // skip creating a list for them here, because they aren't usable
+    // for the splicing mutation (because they are not all interchangable).
+    // But for simplicity we can just leave it as-is for now.
+
```

Comment:
I would still prefer not to add terminal nodes to the chunk store.

Maybe we can modify `node_free_only_self` to free the node itself and its associated terminal children?

Reply:
They *must* be kept alive by the chunk store in some form or another or their parent non-terminal nodes won't have them available. As you noted, freeing them would get tricky if they aren't in `seen_chunks`, and they would also get duplicated for every parent that used them. That might be fairly wasteful of memory even for short terminal nodes, because pointing at an existing node in `seen_chunks` is "free" but creating an entire `node_t` is not. Also short terminal nodes are more likely to be used multiple times (single letters or digits, etc).

One option would be to put them into only into `seen_chunks` and not into `chunk_store`. This would save a little bit of memory and can be accomplished with an `if(node->id != 0)` block that goes from here through the `list_append(node_list, node);` call. I didn't feel like that was worth the complexity when I was making this patch, but I don't object to it either.

Reply:
Note that any nodes used from the `chunk_store` need to be cloned via `node_clone()` before they are used, because their "parent" node pointers are all broken (on purpose) as a result of "collapsing" identical child nodes. AKA, subnodes that are equal will have multiple parents inside the `chunk_store` but the nodes don't have a way to represent that so they just point at whomever their parent was last. We could change it to have "null" for all the parent pointers to make it less confusing.

Anyway, this is why I don't think it is a problem to have the terminal nodes in there as well. They just act the same as other nodes in the `chunk_store` and then get cloned out as proper trees when used.

Reply:
I see. It makes sense.